### ISSUE TITLE
Generalize integration of parameter node with scripted module

### ISSUE THREAD

**Issue by jcfr** [Slicer/Slicer#6724]
For the sake of standardization, I wonder if it would be appropriate to add some abstract classes that use this, and make the extension wizard use that abstract class. If not an abc or mixin, I think it would be good to reduce the amount of boilerplate needed to correctly use parameter nodes (preferably wrapped parameter nodes). I'd hope putting that mechanism in the extension wizard or other templates would help encourage module developers to use the same patterns and make static analysis more viable.

A rough idea of the API might be:

```python
class VectorToScalarVolume(SimpleScriptedModule):
    UI_RESOURCE_PATH = 'UI/VectorToScalarVolume.ui'
    PARAMETER_NODE_TYPE = VectorToScalarVolumeParameterNode

    def __init__(self, parent):
        super().__init__(parent)

        self.logic = None

    def setup(self):
        super().setup()  # automatically sets up `self.params`

        self.logic = VectorToScalarVolumeLogic()

        self.ui.inputSelector.currentNodeChanged.connect(self.onGUIChanged)

        ...

    def initializeParameterNode(self):  # no boilerplate; called by the base
        if not self.params.InputVolume:
            ...

    def updateGUIFromParameterNode(self):  # no boilerplate
        self.ui.inputSelector.setCurrentNode(self.params.InputVolume)
        ...

    def updateParameterNodeFromGUI(self):  # no boilerplate
        self.params.InputVolume = self.ui.inputSelector.currentNode

    def onApplyButton(self):
        ...
```

The main idea is there's no boilerplate for managing the parameter node, avoiding recursive gui/parameter updates, etc.

I'm not sure what exactly the scope of this kind of feature should be, or whether a base class is the right mechanism. Maybe a mixin, or method decorators? Either way, I don't think this kind of logic belongs in the existing `ScriptedLoadableModule` classes; some modules customize the workflow significantly and this would just get in the way.

---

I'll think on it more and wait for some preliminary feedback here... if there seems to be a reasonable way forward I'll open a new issue and try prototyping something.

At the very least, updating the extension wizard's output to use `@parameterNodeWrapper` would be a start.

_Originally posted by @allemangD in https://github.com/Slicer/Slicer/issues/6684#issuecomment-1341627335_

**Comment by allemangD** [comment]
Thanks for opening the issue @jcfr. I have been thinking about it a bit more, I think a base class is _not_ the right solution. I also discovered that the `PARAMETER_NODE_TYPE` pattern does not support static analysis, so I'd prefer something else.

Perhaps a generic mixin like `ParameterNodeBackedGui[VectorToVectorScalarVolumeParameterNode]`? Seems too wordy.

---

To try to focus the scope of this issue, here are the particular patterns I think are worth generalizing:

Building module widgets

- Parse a `.ui` file and generate the `self.ui` namespace
- Build a widget in a method, and assign widgets to `self`. Could encourage a `self.ui` namespace here, too.

Synchronize widget values with parameter node parameters

- Avoid recursion in observations and slots
- Use `slicer.util.NodeModify`
- Use correct nodes on scene change, or when called by other modules.
- Support changing the parameter set via combo box
- A declarative way to set which widgets propagate changes to the parameter node and vice versa.

Static analysis and type hints

- Static analysis in the `self.ui` namespace would be nice.
- Encourage typed parameter node and type hints in `ScriptedLoadableModuleLogic` classes.
- Automatically wrap `getParameterNode` result?

**Comment by lassoan** [comment]
I agree that there is too much boilerplate code in the module template. Many things could be pushed up into the scripted module base classes or maybe in a mixin. But probably simply adding things into the base classes directly would be simpler.

Note that there is already some useful code in slicer.util for automatically synchronizing parameter values and GUI widgets:

https://github.com/Slicer/Slicer/blob/main/Base/Python/slicer/util.py#L386-L511

**Comment by pieper** [comment]
+1 for a more streamlined module template.  I used to create one-off modules all the time when doing slicer experiments, but lately I find there's too much noise, but things break when I try to strip it away.  Instead I end up writing stand-along scripts but those are not as easily converted into deliverable code.

**Comment by Connor-Bowley** [comment]
I am working on adding support for the parameter node wrapper to integrate with a set of widget types.

Current list of widgets I am planning to support:

- QCheckBox -> bool
- QSpinBox -> int
- QDoubleSpinBox -> float
- QComboBox -> (int, float, str, bool) w/Choice annotation
- QLineEdit -> str
- QTextEdit -> str
- ctkPathLineEdit -> pathlib.Path
- ctkDirectoryButton -> pathlib.Path
- qMRMLNodeComboBox -> mrmlNode
    - If you specify the specific node type, it will limit to that node type.
    - If you specify something like `typing.Union[vtkMRMLModelNode, vtkMRMLVolumeNode]` it will limit to those node types
- QSlider(s) -> int
- ctkSliderWidget -> float
- QRadioButton -> bool
- qMRMLTreeView/qMRMLSubjectHierarchyTreeView -> mrmlNode (maybe also support list\[mrmlNode]?)
- QPushButton (checkable) -> bool

The spinbox/slider widgets will recognize the Minimum, Maximum, and WithinRange validators and update the widget accordingly.

Are there any other widgets that should be supported?
